Migrate filesystem handlers to use new resource implementation (ref-point migration 2/10)

tobous opened the pull request:
Migrates the filesystem handler logic of Saros/E to use the new resource model introduced in #1023. Outdated javadoc and method names that are not directly related to the changed logic will be updated in a separate clean-up PR at the end.

### Reviewing this PR

I would suggest reviewing this PR commit by commit.

### Commits

<details><summary><b>[INTERNAL][E] Migrate ResourceDeltaVisitor to new resource impls</b></summary>
<br>

Adjusts the logic to check for reference points instead of projects.

</details>

<details><summary><b>[INTERNAL][E] Skip non-shared resources for ResourceDeltaVisitor</b></summary>
<br>

Adjusts ResourceChangeValidator to only check resource deltas for the
shared reference points. This avoid iterating all non-shared resources.

Adjusts ResourceDeltaVisitor to get passed the reference point to use.
This skips guessing which reference point the resource could belong to.

</details>

<details><summary><b>[INTERNAL][E] Improve logic to ignore deltas caused by opening projects </b></summary>
<br>

Adjusts the logic in SharedResourcesManager used to determine whether a
resource delta was caused by opening a project. Instead of manually
tracking the state of shared projects, the new filter instead checks
whether the delta describes a project being opened.

Removes the tracking logic for open shared projects.

</details>

<details><summary><b>[INTERNAL][E] Skip non-shared resources for ProjectDeltaVisitor</b></summary>
<br>

Adjusts SharedResourcesManager to only call ProjectDeltaVisitor on
deltas for the shared reference points. This avoids a lot of overhead
otherwise caused by iterating and converting all non-shared changed
resources.

Adjusts ProjectDeltaVisitor to get passed a reference point object to
use. This skips having to guess which reference point the resources
belong to.

Removes ProjectDeltaVisitor.reset() as there now is one visitor per
reference point, meaning it does not have to get reset.

Adds the utility method createProjectReferencePointMapping which creates
a map of projects onto their contained reference points. This is useful
to quickly access all reference points for a specific project. This is
needed as we still filter by projects to check whether the resource
delta was created by the project being opened.

</details>

<b>[INTERNAL][E] Migrate ProjectDeltaVisitor to new resource impls</b>

<details><summary><b>[INTERNAL][E] Check if resources are shared in ProjectDeltaVisitor</b></summary>
<br>

Adds a check whether the changed resources are actually shared to the
ProjectDeltaVisitor. Even though it specifically only deals with deltas
affecting resources below a shared reference point, there might still be
resources that are not part of the session as they are seen as ignored.
This is defined through IResource.isIgnored().

Adjusts the move logic to create partial activities if the source or the
target file is not shared. If the source is not shared, a creation
activity is dispatched instead of the move activity. If the target is
not shared, a deletion activity is dispatched instead of the move
activity.

</details>

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java  1
- eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java  2
- eclipse/src/saros/project/SharedResourcesManager.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5736374&bid=18764446)

srossbach commented on eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java:
```diff
@@ -60,22 +64,34 @@ public IStatus validateChange(final IResourceDelta delta, final IProgressMonitor
+
+      IResourceDelta referencePointDelta = delta.findMember(referencePointDelegate.getFullPath());
+
+      final ResourceDeltaVisitor visitor = new ResourceDeltaVisitor(referencePoint, currentSession);
```
Should be place above line 80, not here.

tobous replied:
Will do.

tobous replied:
Done.

srossbach commented on eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java:
```diff
@@ -60,22 +64,34 @@ public IStatus validateChange(final IResourceDelta delta, final IProgressMonitor
 
-    if (!currentSession.hasWriteAccess() && visitor.isModifyingResources())
-      return MODIFYING_RESOURCES_ERROR_STATUS;
+      if (referencePointDelta != null) {
```
From the documentation:

The supplied path may be absolute or relative; in either case, it is interpreted as relative to this delta. Trailing separators are ignored. If the path is empty this delta is returned. 

So is /Foo/Bar/Foo (fullpath) applied on the Delta (/Foo/Bar) checked as /Foo/Bar/Foo or /Foo/Bar/Foo/Bar/Foo ?

tobous replied:
The way I understand it, the initial delta always start at the workspace root. So, as we are always working on the base delta here, using the full path should result in the correct directory.

But I will check if this assumption holds true. If it does not, I can adjust the logic to get the resource of the delta, get its full path and then compute the relative path between the base resource and the reference point resource (if there is one).

tobous replied:
From my testing, the initial delta always starts at the workspace root. Any non-changed resources are marked as such in the delta. But I could not find any official documentation on this behavior.

So if you prefer, I can still add the logic to create the relative path between the delta resource and the reference point delegate and use it instead anyways just to make sure.

tobous commented on eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java:
```diff
@@ -60,22 +64,34 @@ public IStatus validateChange(final IResourceDelta delta, final IProgressMonitor
 
-    if (visitor.isDeletingProject()) return DELETE_PROJECT_ERROR_STATUS;
+      try {
+        delta.accept(visitor);
```
Should by `referencePointDelta`!

tobous replied:
Done.

tobous commented:
Rebased onto current base branch without any interaction.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java  1
- eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java  2
- eclipse/src/saros/project/SharedResourcesManager.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5736374&bid=18764446)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java  1
- eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java  2
- eclipse/src/saros/project/SharedResourcesManager.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5736374&bid=18764446)

tobous commented on eclipse/src/saros/project/SharedResourcesManager.java:
```diff
@@ -223,39 +199,59 @@ private void handlePostChange(IResourceChangeEvent event) {
       }
 
-      resourceActivities.addAll(projectDeltaVisitor.getActivities());
+      for (IReferencePoint referencePoint : projectReferencePointMap.get(project)) {
```
```suggestion
      for (IReferencePoint referencePoint : projectReferencePoints) {
```

srossbach commented on eclipse/src/saros/project/SharedResourcesManager.java:
```diff
@@ -83,30 +84,7 @@
-      new ISessionListener() {
-
-        @Override
-        public void referencePointAdded(IReferencePoint referencePoint) {
```
Any reason to not use this listeners to update a map that you just calculate over and over again in createProjectReferencePointMap ?

tobous replied:
I would like to avoid caching this mapping as updating it is not always that easy. I am not sure how Eclipse handles internal renamings/moves (i.e. whether the existing resource objects for the changed resources are updated or whether they are just invalidated).

Furthermore, a user could theoretically move a reference point to a different project. (I know this is currently not strictly supported by Saros/E, but still. Also, I don't see a reason why to not support this case. With the new reference point implementation, renaming/moving a reference point or one of its parents should not be an issue if resource objects are updated internally for such changes. This is already supported by Saros/I.)

Also, since this is only done once per resource delta, creating the mapping in O(#ref-points) is acceptable in my opinion.

srossbach commented on eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java:
```diff
@@ -32,17 +38,22 @@ public boolean visit(final IResourceDelta delta) throws CoreException {
+    saros.filesystem.IResource resourceWrapper =
+        ResourceConverter.convertToResource(referencePoint, resource);
+
+    if (resourceWrapper == null || !session.isShared(resourceWrapper)) return true;
```
Isn't there a null check anymore in the isShared method ?

tobous replied:
Yes, `isShared` still checks for `null`. I am kind of inconsistent on how to handle this. I think for most calls in Saros/I, I still explicitly check for `null` (in case the internal implementation changes?). But I don't think it matter that much. If you want, I can still remove it.

tobous commented on eclipse/src/saros/project/ProjectDeltaVisitor.java:
```diff
@@ -170,7 +164,18 @@ private void handleFileDelta(IResourceDelta delta) {
+        ResourceConverter.convertToResource(referencePoint, resource);
+
+    if (wrappedResource == null) {
+      log.error(
```
I am not quite sure what you mean by "unsharing".

If you mean "removing a reference point from the session": I am not sure whether reducing the resource scope was ever considered for the session design. This would probably also lead to a lot of other issues.

If you mean "excluding specific child resources of a reference point from the session": Not necessarily. The current logic operates solely on paths, so it would still return resources on such ignored/excluded children. This is, however, missing a check to not send activities for such files. I will add an `isShared` check.

srossbach replied:
Well there is a referencePointRemove method, but yeah you are right. If you really try to implement this things will get worse and you will find a lot of design issues / choice that will render that to almost unable to implement it properly.

tobous commented on eclipse/src/saros/project/ProjectDeltaVisitor.java:
```diff
@@ -291,9 +339,27 @@ private void addActivity(IResourceActivity<? extends saros.filesystem.IResource>
+   * @param fullResourcePath the full resource path to check
+   * @return whether the given full resource path points to a child resource of the reference point
+   */
+  private boolean isSameReferencePoint(IPath fullResourcePath) {
```
Will do.

tobous commented on eclipse/src/saros/project/ProjectDeltaVisitor.java:
```diff
@@ -291,9 +339,27 @@ private void addActivity(IResourceActivity<? extends saros.filesystem.IResource>
+   * @return the reference-point-relative path representing the given full path or the given full
+   *     path if no such path could be constructed
+   */
+  private IPath getReferencePointRelativePath(IPath fullResourcePath) {
```
Will do.

tobous commented:
Applied review comments. Added `isShared` checks before creating the activities in `ProjectDeltaVisitor`. Adjusted `generateMoved` to create a deletion activity if only the source file is shared and a creation activity if only the target file is shared.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java  1
- eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java  2
- eclipse/src/saros/project/SharedResourcesManager.java  2
- eclipse/src/saros/project/ProjectDeltaVisitor.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5736374&bid=18764446)

tobous commented:
Rebased onto current base branch without any interaction.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java  1
- eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java  2
- eclipse/src/saros/project/SharedResourcesManager.java  2
- eclipse/src/saros/project/ProjectDeltaVisitor.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5736374&bid=18764446)

tobous commented:
Rebased onto current base branch without any interactions.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java  1
- eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java  2
- eclipse/src/saros/project/SharedResourcesManager.java  2
- eclipse/src/saros/project/ProjectDeltaVisitor.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5736374&bid=18764446)

tobous commented:
Rebased onto current base branch without any interaction.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java  1
- eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java  2
- eclipse/src/saros/project/SharedResourcesManager.java  2
- eclipse/src/saros/project/ProjectDeltaVisitor.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5736374&bid=18764446)

tobous commented:
Rebased onto current master without any interaction.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- eclipse/src/saros/session/resources/validation/ResourceDeltaVisitor.java  1
- eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java  2
- eclipse/src/saros/project/SharedResourcesManager.java  2
- eclipse/src/saros/project/ProjectDeltaVisitor.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5736374&bid=18764446)
